diff -p -up MigrationTools-45/migrate_all_online.sh.suffix MigrationTools-45/migrate_all_online.sh
--- MigrationTools-45/migrate_all_online.sh.suffix	2009-04-01 09:11:59.000000000 +0200
+++ MigrationTools-45/migrate_all_online.sh	2009-04-01 09:11:59.000000000 +0200
@@ -192,6 +192,31 @@ $PERL -I${INSTDIR} ${INSTDIR}migrate_net
 echo "Migrating netgroups (by host)..."
 $PERL -I${INSTDIR} ${INSTDIR}migrate_netgroup_byhost.pl	$ETC_NETGROUP >> $DB
 
+# Try to add the top-level object, because it's usually missing.
+TYPE=`echo $LDAP_BASEDN | cut -f1 -d=`
+TOP=`echo $LDAP_BASEDN | cut -f1 -d, | cut -f2 -d=`
+CLASS=$TYPE
+if [ "$CLASS" = "dc" ] ; then
+	CLASS=domain
+elif [ "$CLASS" = "o" ] ; then
+	CLASS=organization
+elif [ "$CLASS" = "l" ] ; then
+	CLASS=locality
+elif [ "$CLASS" = "c" ] ; then
+	CLASS=country
+elif [ "$CLASS" = "nismapname" ] ; then
+	CLASS=nisMap
+elif [ "$CLASS" = "cn" ] ; then
+	CLASS=container
+fi
+
+cat << EOF | $LDAPADD -h $LDAPHOST -D "$LDAP_BINDDN" -x -w "$LDAP_BINDCRED"
+dn: $LDAP_BASEDN
+$TYPE: $TOP
+objectClass: top
+objectClass: $CLASS
+EOF
+
 echo "Importing into LDAP..."
 
 if [ `basename $LDAPADD` = "slapadd" ]; then
