--- openldap-2.3.34/tests/scripts/all.dont-write-testdir	2007-01-02 23:44:13.000000000 +0200
+++ openldap-2.3.34/tests/scripts/all	2007-03-13 19:30:28.000000000 +0200
@@ -13,7 +13,7 @@
 ## top-level directory of the distribution or, alternatively, at
 ## <http://www.OpenLDAP.org/license.html>.
 
-SHTOOL="$SRCDIR/../build/shtool"
+. $SRCDIR/scripts/defines.sh
 TB=`$SHTOOL echo -e "%B"`
 TN=`$SHTOOL echo -e "%b"`
 SLEEPTIME=10
@@ -23,9 +23,9 @@
 for CMD in $SRCDIR/scripts/test*; do
 	# remove cruft from prior test
 	if test $PRESERVE = yes ; then
-		/bin/rm -rf testrun/db.*
+		/bin/rm -rf $TESTDIR/db.*
 	else
-		/bin/rm -rf testrun
+		/bin/rm -rf $TESTDIR
 	fi
 
 	echo ">>>>> Starting ${TB}`basename $CMD`${TN} ..."
--- openldap-2.3.34/tests/scripts/defines.sh.dont-write-testdir	2007-01-02 23:44:13.000000000 +0200
+++ openldap-2.3.34/tests/scripts/defines.sh	2007-03-13 19:44:04.000000000 +0200
@@ -13,6 +13,8 @@
 ## top-level directory of the distribution or, alternatively, at
 ## <http://www.OpenLDAP.org/license.html>.

+TESTWD=`pwd`
+
 MONITORDB=${AC_monitor-no}
 BACKLDAP=${AC_ldap-ldapno}
 BACKMETA=${AC_meta-metano}
@@ -39,7 +39,7 @@
 
 PROGDIR=./progs
 DATADIR=${USER_DATADIR-./testdata}
-TESTDIR=${USER_TESTDIR-./testrun}
+TESTDIR=${USER_TESTDIR-$TESTWD/testrun}
 SCHEMADIR=${USER_SCHEMADIR-./schema}
 
 DBDIR1A=$TESTDIR/db.1.a
@@ -115,6 +115,7 @@
 CONF5=$TESTDIR/slapd.5.conf
 CONF6=$TESTDIR/slapd.6.conf
 ADDCONF=$TESTDIR/slapadd.conf
+CONFLDIF=$TESTDIR/slapd-dynamic.ldif
 
 TOOLARGS="-x $LDAP_TOOLARGS"
 TOOLPROTO="-P 3"
@@ -298,5 +299,7 @@
 ACIOUT=$DATADIR/aci.out
 DYNLISTOUT=$DATADIR/dynlist.out
 
+SHTOOL="$SRCDIR/../build/shtool"
+
 # Just in case we linked the binaries dynamically
 LD_LIBRARY_PATH=`pwd`/../libraries:${LD_LIBRARY_PATH} export LD_LIBRARY_PATH
--- openldap-2.3.34/tests/scripts/test034-translucent.dont-write-testdir	2007-01-02 23:44:13.000000000 +0200
+++ openldap-2.3.34/tests/scripts/test034-translucent	2007-03-13 19:30:28.000000000 +0200
@@ -38,6 +38,9 @@
 # configure backside
 mkdir -p $TESTDIR $DBDIR1
 
+$SLAPPASSWD -g -n >$CONFIGPWF
+echo "rootpw `$SLAPPASSWD -T $CONFIGPWF`" >$TESTDIR/configpw.conf
+
 . $CONFFILTER $BACKEND $MONITORDB < $TRANSLUCENTREMOTECONF > $CONF1
 echo "Running slapadd to build remote slapd database..."
 $SLAPADD -f $CONF1 -l $LDIFTRANSLUCENTCONFIG
--- openldap-2.3.34/tests/scripts/test042-valsort.dont-write-testdir	2007-01-02 23:44:13.000000000 +0200
+++ openldap-2.3.34/tests/scripts/test042-valsort	2007-03-13 19:30:28.000000000 +0200
@@ -23,6 +23,9 @@
 
 mkdir -p $TESTDIR $DBDIR1
 
+$SLAPPASSWD -g -n >$CONFIGPWF
+echo "rootpw `$SLAPPASSWD -T $CONFIGPWF`" >$TESTDIR/configpw.conf
+
 echo "Running slapadd to build slapd database..."
 . $CONFFILTER $BACKEND $MONITORDB < $VALSORTCONF > $CONF1
 $SLAPADD -f $CONF1 -l $LDIFVALSORT
--- openldap-2.3.34/tests/run.in.dont-write-testdir	2007-01-02 23:44:12.000000000 +0200
+++ openldap-2.3.34/tests/run.in	2007-03-13 19:39:15.000000000 +0200
@@ -172,9 +172,9 @@
 		echo "User data directory (userdata) does not exist."
 		exit 1
 	fi
-	mkdir -p ${TESTDIR}
 	cp -R userdata/* ${TESTDIR}
 fi
+mkdir -p ${TESTDIR}
 
 # disable LDAP initialization
 LDAPNOINIT=true; export LDAPNOINIT
