--- openldap-2.3.43/include/ldap_pvt_thread.h.orig	2008-02-12 00:24:10.000000000 +0100
+++ openldap-2.3.43/include/ldap_pvt_thread.h	2009-10-12 13:35:42.000000000 +0200
@@ -57,12 +57,12 @@
 
 #ifndef LDAP_PVT_THREAD_H_DONE
 #define	LDAP_PVT_THREAD_SET_STACK_SIZE
-#ifndef LDAP_PVT_THREAD_STACK_SIZE
-	/* LARGE stack. Will be twice as large on 64 bit machine. */
-#define LDAP_PVT_THREAD_STACK_SIZE	( 1 * 1024 * 1024 * sizeof(void *) )
-/* May be explicitly defined to zero to disable it */
-#elif LDAP_PVT_THREAD_STACK_SIZE == 0
-#undef LDAP_PVT_THREAD_SET_STACK_SIZE
+/* The size may be explicitly #defined to zero to disable it. */
+#if defined( LDAP_PVT_THREAD_STACK_SIZE ) && LDAP_PVT_THREAD_STACK_SIZE == 0
+#       undef LDAP_PVT_THREAD_SET_STACK_SIZE
+#elif !defined( LDAP_PVT_THREAD_STACK_SIZE )
+        /* LARGE stack. Will be twice as large on 64 bit machine. */
+#       define LDAP_PVT_THREAD_STACK_SIZE ( 1 * 1024 * 1024 * sizeof(void *) )
 #endif
 #endif /* !LDAP_PVT_THREAD_H_DONE */
 
--- openldap-2.3.43/libraries/libldap/os-ip.c~	2008-05-20 01:28:54.000000000 +0200
+++ openldap-2.3.43/libraries/libldap/os-ip.c	2009-10-12 11:49:03.000000000 +0200
@@ -652,7 +652,7 @@ ldap_host_connected_to( Sockbuf *sb, con
 		char *herr;
 #ifdef NI_MAXHOST
 		char hbuf[NI_MAXHOST];
-#elif defined( MAXHOSTNAMELEN
+#elif defined MAXHOSTNAMELEN
 		char hbuf[MAXHOSTNAMELEN];
 #else
 		char hbuf[256];
