--- smbk5pwd.c~	2008-02-12 00:34:15.000000000 +0100
+++ smbk5pwd.c	2008-06-25 16:51:44.000000000 +0200
@@ -350,6 +350,7 @@
 	Modifications *ml;
 	slap_overinst *on = (slap_overinst *)op->o_bd->bd_info;
 	smbk5pwd_t *pi = on->on_bi.bi_private;
+    char term;
 
 	/* Not the operation we expected, pass it on... */
 	if ( ber_bvcmp( &slap_EXOP_MODIFY_PASSWD, &op->ore_reqoid ) ) {
@@ -360,6 +361,9 @@
 	rc = be_entry_get_rw( op, &op->o_req_ndn, NULL, NULL, 0, &e );
 	if ( rc != LDAP_SUCCESS ) return rc;
 
+    term = qpw->rs_new.bv_val[qpw->rs_new.bv_len];
+    qpw->rs_new.bv_val[qpw->rs_new.bv_len] = '\0';
+
 #ifdef DO_KRB5
 	/* Kerberos stuff */
 	do {
@@ -596,6 +600,7 @@
 	}
 #endif /* DO_SAMBA */
 	be_entry_release_r( op, e );
+    qpw->rs_new.bv_val[qpw->rs_new.bv_len] = term;
 
 	return SLAP_CB_CONTINUE;
 }
