diff -p -up openldap-2.3.43/servers/slapd/slap.h.cldap openldap-2.3.43/servers/slapd/slap.h
--- openldap-2.3.43/servers/slapd/slap.h.cldap	2008-02-12 00:24:18.000000000 +0100
+++ openldap-2.3.43/servers/slapd/slap.h	2009-04-01 08:55:56.000000000 +0200
@@ -33,6 +33,7 @@
 #include <ac/stdlib.h>
 
 #include <sys/types.h>
+#include <sys/socket.h>
 #include <ac/syslog.h>
 #include <ac/regex.h>
 #include <ac/signal.h>
