--- openldap-2.3.43/servers/slapd/schema/misc.schema.mdk	2008-02-12 00:24:26.000000000 +0100
+++ openldap-2.3.43/servers/slapd/schema/misc.schema	2009-04-01 09:18:06.000000000 +0200
@@ -25,7 +25,7 @@
 #	(a work in progress)
 #
 attributetype ( 2.16.840.1.113730.3.1.13
-	NAME 'mailLocalAddress'
+	NAME ('mailLocalAddress' 'mailAlternateAddress')
 	DESC 'RFC822 email address of this recipient'
 	EQUALITY caseIgnoreIA5Match
 	SYNTAX 1.3.6.1.4.1.1466.115.121.1.26{256} )
@@ -38,7 +38,7 @@ attributetype ( 2.16.840.1.113730.3.1.18
 	SINGLE-VALUE )
 
 attributetype ( 2.16.840.1.113730.3.1.47
-	NAME 'mailRoutingAddress'
+	NAME ('mailForwardingAddress' 'mailRoutingAddress')
 	DESC 'RFC822 routing address of this recipient'
 	EQUALITY caseIgnoreIA5Match
 	SYNTAX 1.3.6.1.4.1.1466.115.121.1.26{256}
@@ -73,3 +73,10 @@ objectclass ( 1.3.6.1.4.1.42.2.27.1.2.5
 	SUP top STRUCTURAL
 	MUST cn
 	MAY rfc822MailMember )
+
+objectclass ( 2.16.840.1.113730.3.2.3
+	NAME 'mailRecipient'
+	DESC 'Internet local mail recipient'
+	SUP top AUXILIARY
+	MAY     ( cn $ mail $ mailAlternateAddress $ mailHost $ mailRoutingAddress ) )
+
